File: python/zamlet/lane_order.py
from enum import IntEnum

from zamlet.moore import moore_d2xy, moore_xy2d


class LaneOrder(IntEnum):
    MOORE = 0
    UNKNOWN1 = 1
    ROW_MAJOR = 2
    TOROIDAL_ROW_MAJOR = 3
    COLUMN_MAJOR = 4
    TOROIDAL_COLUMN_MAJOR = 5

    @classmethod
    def count(cls) -> int:
        return len(cls)


def vw_index_to_j_coords(params, lane_order: LaneOrder, vw_index: int):
    if lane_order == LaneOrder.ROW_MAJOR:
        jx = vw_index % (params.j_cols * params.k_cols)
        jy = vw_index // (params.j_cols * params.k_cols)
        return jx, jy
    if lane_order == LaneOrder.MOORE:
        total_cols = params.j_cols * params.k_cols
        total_rows = params.j_rows * params.k_rows
        assert total_cols == total_rows, (
            f"MOORE requires square grid, got {total_cols}x{total_rows}"
        )
        assert total_cols & (total_cols - 1) == 0, (
            f"MOORE requires power-of-2 grid, got {total_cols}"
        )
        return moore_d2xy(total_cols, vw_index)
    if lane_order == LaneOrder.COLUMN_MAJOR:
        total_rows = params.j_rows * params.k_rows
        jx = vw_index // total_rows
        jy = vw_index % total_rows
        return jx, jy
    if lane_order == LaneOrder.TOROIDAL_ROW_MAJOR:
        total_cols = params.j_cols * params.k_cols
        total_rows = params.j_rows * params.k_rows
        assert total_rows >= 2
        row_rank = vw_index // total_cols
        ordered_x = vw_index % total_cols
        if row_rank < (total_rows + 1) // 2:
            y = row_rank * 2
        else:
            y = 2 * (total_rows - 1 - row_rank) + 1
        x = total_cols - 1 - ordered_x if row_rank % 2 else ordered_x
        return x, y
    if lane_order == LaneOrder.TOROIDAL_COLUMN_MAJOR:
        total_cols = params.j_cols * params.k_cols
        total_rows = params.j_rows * params.k_rows
        assert total_cols >= 2
        col_rank = vw_index // total_rows
        ordered_y = vw_index % total_rows
        if col_rank < (total_cols + 1) // 2:
            x = col_rank * 2
        else:
            x = 2 * (total_cols - 1 - col_rank) + 1
        y = total_rows - 1 - ordered_y if col_rank % 2 else ordered_y
        return x, y
    raise NotImplementedError(f"Lane order {lane_order}")


def vw_index_to_routing_coords(params, lane_order: LaneOrder, vw_index: int):
    jx, jy = vw_index_to_j_coords(params, lane_order, vw_index)
    return jx + params.west_offset, jy + params.north_offset


def j_coords_to_vw_index(params, lane_order: LaneOrder, jx: int, jy: int):
    if lane_order == LaneOrder.ROW_MAJOR:
        return jy * (params.j_cols * params.k_cols) + jx
    if lane_order == LaneOrder.MOORE:
        total_cols = params.j_cols * params.k_cols
        total_rows = params.j_rows * params.k_rows
        assert total_cols == total_rows, (
            f"MOORE requires square grid, got {total_cols}x{total_rows}"
        )
        assert total_cols & (total_cols - 1) == 0, (
            f"MOORE requires power-of-2 grid, got {total_cols}"
        )
        return moore_xy2d(total_cols, jx, jy)
    if lane_order == LaneOrder.COLUMN_MAJOR:
        return jx * (params.j_rows * params.k_rows) + jy
    if lane_order == LaneOrder.TOROIDAL_ROW_MAJOR:
        total_cols = params.j_cols * params.k_cols
        row_pair = jy >> 1
        row_rank = (params.j_rows * params.k_rows - 1) - row_pair if jy & 1 else row_pair
        ordered_x = (total_cols - 1) - jx if row_rank & 1 else jx
        return row_rank * total_cols + ordered_x
    if lane_order == LaneOrder.TOROIDAL_COLUMN_MAJOR:
        total_rows = params.j_rows * params.k_rows
        col_pair = jx >> 1
        col_rank = (params.j_cols * params.k_cols - 1) - col_pair if jx & 1 else col_pair
        ordered_y = (total_rows - 1) - jy if col_rank & 1 else jy
        return col_rank * total_rows + ordered_y
    raise NotImplementedError(f"Lane order {lane_order}")


def vw_index_to_k_indices(params, lane_order: LaneOrder, vw_index: int):
    jx, jy = vw_index_to_j_coords(params, lane_order, vw_index)
    kx = jx // params.j_cols
    ky = jy // params.j_rows
    k_index = ky * params.k_cols + kx
    j_in_k_x = jx % params.j_cols
    j_in_k_y = jy % params.j_rows
    j_in_k_index = j_in_k_y * params.j_cols + j_in_k_x
    return k_index, j_in_k_index


def k_indices_to_j_coords(params, k_index: int, j_in_k_index: int):
    kx = k_index % params.k_cols
    ky = k_index // params.k_cols
    j_in_k_x = j_in_k_index % params.j_cols
    j_in_k_y = j_in_k_index // params.j_cols
    jx = kx * params.j_cols + j_in_k_x
    jy = ky * params.j_rows + j_in_k_y
    return jx, jy


def k_indices_to_routing_coords(params, k_index: int, j_in_k_index: int):
    jx, jy = k_indices_to_j_coords(params, k_index, j_in_k_index)
    return jx + params.west_offset, jy + params.north_offset


def k_indices_to_vw_index(
    params, lane_order: LaneOrder, k_index: int, j_in_k_index: int,
):
    jx, jy = k_indices_to_j_coords(params, k_index, j_in_k_index)
    return j_coords_to_vw_index(params, lane_order, jx, jy)

File: src/main/scala/zamlet/LaneOrder.scala
package zamlet

import chisel3._

object LaneOrder extends ChiselEnum {
  val MOORE = Value(0.U)
  val UNKNOWN1 = Value(1.U)
  val ROW_MAJOR = Value(2.U)
  val TOROIDAL_ROW_MAJOR = Value(3.U)
  val COLUMN_MAJOR = Value(4.U)
  val TOROIDAL_COLUMN_MAJOR = Value(5.U)

  def count: Int = all.length
}

object LaneOrderMapping {
  private object MooreOrientation extends ChiselEnum {
    val N = Value(0.U)
    val E = Value(1.U)
    val S = Value(2.U)
    val W = Value(3.U)
  }

  private object MooreQuadrant extends ChiselEnum {
    val NE = Value(0.U)
    val SE = Value(1.U)
    val SW = Value(2.U)
    val NW = Value(3.U)
  }

  private class MooreNode extends Bundle {
    val orientation = MooreOrientation()
    val dir = Bool()
  }

  private def index(order: LaneOrder.Type): Int = order.asUInt.litValue.toInt

  private def qIntoOrientation(
    quadrant: MooreQuadrant.Type,
    orientation: MooreOrientation.Type
  ): MooreQuadrant.Type = {
    (quadrant.asUInt - orientation.asUInt).asTypeOf(quadrant)
  }

  private def oOutOfOrientation(
    original: MooreOrientation.Type,
    orientation: MooreOrientation.Type
  ): MooreOrientation.Type = {
    (original.asUInt + orientation.asUInt).asTypeOf(original)
  }

  private def xyToQuadrant(x: Bool, y: Bool): MooreQuadrant.Type = {
    val q = Wire(MooreQuadrant())
    when (!x && !y) {
      q := MooreQuadrant.NW
    } .elsewhen (!x && y) {
      q := MooreQuadrant.SW
    } .elsewhen (x && !y) {
      q := MooreQuadrant.NE
    } .otherwise {
      q := MooreQuadrant.SE
    }
    q
  }

  private def getSubNode(iNode: MooreNode, quadrant: MooreQuadrant.Type): MooreNode = {
    val oNode = Wire(new MooreNode)
    when (quadrant === MooreQuadrant.NW) {
      oNode.dir := !iNode.dir
      oNode.orientation := oOutOfOrientation(iNode.orientation, MooreOrientation.W)
    } .elsewhen (quadrant === MooreQuadrant.SW) {
      oNode.dir := iNode.dir
      oNode.orientation := iNode.orientation
    } .elsewhen (quadrant === MooreQuadrant.SE) {
      oNode.dir := iNode.dir
      oNode.orientation := iNode.orientation
    } .otherwise {
      oNode.dir := !iNode.dir
      oNode.orientation := oOutOfOrientation(iNode.orientation, MooreOrientation.E)
    }
    oNode
  }

  private def getFirstNode(quadrant: MooreQuadrant.Type): MooreNode = {
    val oNode = Wire(new MooreNode)
    oNode.dir := false.B
    when (quadrant === MooreQuadrant.SE) {
      oNode.orientation := MooreOrientation.W
    } .elsewhen (quadrant === MooreQuadrant.NE) {
      oNode.orientation := MooreOrientation.W
    } .otherwise {
      oNode.orientation := MooreOrientation.E
    }
    oNode
  }

  def rowMajor(params: ZamletParams, x: UInt, y: UInt): UInt = {
    (y * params.jTotalCols.U + x)(params.log2JInL - 1, 0)
  }

  def columnMajor(params: ZamletParams, x: UInt, y: UInt): UInt = {
    (x * params.jTotalRows.U + y)(params.log2JInL - 1, 0)
  }

  def toroidalRowMajor(params: ZamletParams, x: UInt, y: UInt): UInt = {
    require(params.jTotalRows >= 2, "Toroidal row-major lane order requires at least two rows")
    val rowPair = y >> 1
    val rowRank = Mux(y(0), (params.jTotalRows - 1).U - rowPair, rowPair)
    val orderedX = Mux(rowRank(0), (params.jTotalCols - 1).U - x, x)
    (rowRank * params.jTotalCols.U + orderedX)(params.log2JInL - 1, 0)
  }

  def toroidalColumnMajor(params: ZamletParams, x: UInt, y: UInt): UInt = {
    require(params.jTotalCols >= 2, "Toroidal column-major lane order requires at least two columns")
    val columnPair = x >> 1
    val columnRank = Mux(x(0), (params.jTotalCols - 1).U - columnPair, columnPair)
    val orderedY = Mux(columnRank(0), (params.jTotalRows - 1).U - y, y)
    (columnRank * params.jTotalRows.U + orderedY)(params.log2JInL - 1, 0)
  }

  def moore(params: ZamletParams, x: UInt, y: UInt): UInt = {
    require(params.jTotalCols == params.jTotalRows,
      s"Moore lane order requires a square jamlet grid, got ${params.jTotalCols}x${params.jTotalRows}")
    require(params.xPosWidth == params.yPosWidth)

    val nLayers = params.log2JTotal / 2
    val outputNodes = Seq.fill(nLayers - 1)(Wire(new MooreNode))
    val outputIndices = Seq.fill(nLayers)(Wire(UInt((params.xPosWidth + params.yPosWidth).W)))
    val x0 = x(params.log2JTotalCols - 1)
    val y0 = y(params.log2JTotalRows - 1)
    val quadrant = xyToQuadrant(x0, y0)
    outputIndices(0) := quadrant.asUInt
    if (nLayers > 1) {
      outputNodes(0) := getFirstNode(quadrant)
    }

    for (layerIndex <- 1 until nLayers) {
      val xi = x(params.log2JTotalCols - layerIndex - 1)
      val yi = y(params.log2JTotalRows - layerIndex - 1)
      val quadrant = xyToQuadrant(xi, yi)
      val lastNode = outputNodes(layerIndex - 1)
      val rotatedQuadrant = qIntoOrientation(quadrant, lastNode.orientation)
      val index = Wire(UInt(2.W))
      when (lastNode.dir) {
        index := 3.U - rotatedQuadrant.asUInt
      } .otherwise {
        index := rotatedQuadrant.asUInt
      }
      outputIndices(layerIndex) := (outputIndices(layerIndex - 1) << 2) + index
      if (layerIndex < nLayers - 1) {
        outputNodes(layerIndex) := getSubNode(lastNode, rotatedQuadrant)
      }
    }
    outputIndices(nLayers - 1)(params.log2JInL - 1, 0)
  }

  def indices(params: ZamletParams, x: UInt, y: UInt): Vec[UInt] = {
    val result = Wire(Vec(LaneOrder.count, UInt(params.log2JInL.W)))
    for (i <- 0 until LaneOrder.count) {
      result(i) := 0.U
    }
    result(index(LaneOrder.MOORE)) := moore(params, x, y)
    result(index(LaneOrder.ROW_MAJOR)) := rowMajor(params, x, y)
    result(index(LaneOrder.TOROIDAL_ROW_MAJOR)) := toroidalRowMajor(params, x, y)
    result(index(LaneOrder.COLUMN_MAJOR)) := columnMajor(params, x, y)
    result(index(LaneOrder.TOROIDAL_COLUMN_MAJOR)) := toroidalColumnMajor(params, x, y)
    result
  }
}
